Add DPI Restriction switch to UniFi integration

hunterjm opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Funny backstory... We recently caught my daughter watching YouTube on her Chromebook (daily for the past month) while supposed to be in school. In order to put a stop to it, I created a DPI filter in UniFi to block streaming media on the VLAN her machine is connected to.

This PR exposes that as a switch to Home Assistant, allowing the user to enable/disable the restriction without going layers deep in the UniFi settings. More importantly, one can create automations to time-box the restriction which is not a part of the standard UniFi functionality.

**Note: This PR is dependent on a PR in the parent library:** https://github.com/Kane610/aiounifi/pull/44

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: N/A
- This PR is related to issue: N/A
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/15489

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [x] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

hunterjm committed:
initial implementation for controlling DPI restrictions

Kane610 reviewed (changes requested):
Overall looks good!

hunterjm committed:
address PR review comments and add DataUpdateCoordinator

hunterjm committed:
fix existing tests against new lib version

hunterjm commented:
Comments addressed and existing tests fixed once lib is bumped.
![image](https://user-images.githubusercontent.com/747670/97522558-69651680-1976-11eb-8492-759a3e378b29.png)

hunterjm committed:
add tests for DPI switches

hunterjm commented:
@Kane610 - Tests added for DPI switches and passes locally. We are now only blocked on a new release of `aiounifi` containing https://github.com/Kane610/aiounifi/pull/44

Kane610 commented:
> @Kane610 - Tests added for DPI switches and passes locally. We are now only blocked on a new release of `aiounifi` containing [Kane610/aiounifi#44](https://github.com/Kane610/aiounifi/pull/44)

v24 is releasing right now.

hunterjm committed:
bump aiounifi

Kane610 commented on homeassistant/components/unifi/switch.py:
```diff
@@ -57,9 +69,28 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
             controller.api.clients.process_raw([client.raw])
 
+    # Create update coordinator for DPI Restrictions
+    if controller.option_dpi_restrictions:
```
This will not work if the user enables it after everything has loaded. You need to handle loading/unloading of this from the items_added function

hunterjm replied:
Since Update Coordinator refresh won't be called unless there are subscribers, I think I can remove this conditional and always create the coordinator.

Kane610 commented on homeassistant/components/unifi/switch.py:
```diff
@@ -284,3 +338,73 @@ async def options_updated(self) -> None:
+        UniFiBase.__init__(self, item, controller)
+
+    @property
+    def key(self) -> Any:
```
Shouldn't it always be a string?
```suggestion
    def key(self) -> str:
```

hunterjm replied:
Today with `mac`, `_id`, and `name` as the only keys, it is. But there's nothing really that says it **needs** to be a string. If later we use one of the non-string attributes as the key.

Kane610 replied:
Ok

Kane610 reviewed (changes requested):
Nearly there!

There are no web socket events at all for DPI?

hunterjm commented:
> There are no web socket events at all for DPI?

I just looked into this. It appears there are `dpiapp:{add/remove/sync}` and `dpigroup:{add/remove/sync}` events.

Kane610 commented:
> > There are no web socket events at all for DPI?
> 
> I just looked into this. It appears there are `dpiapp:{add/delete/sync}` and `dpigroup:{add/delete/sync}` websocket messages. Not sure how to implement handling these though.

You figure it out :), if you solved xbox this is a piece of cake ;)
https://github.com/Kane610/aiounifi/blob/9a4f774f100f1e11ec884402b8ee08c9d705dc63/aiounifi/controller.py#L157

hunterjm commented:
The main problem is that I am combining the context of groups and apps to make this more simple for Home Assistant. With the events being different for group/app it makes message handling impossible, because I need more context than provided in a single websocket event, so would have to poll the API anyway without a somewhat large refactor of the parent library.

hunterjm committed:
listen to events instead of polling

hunterjm committed:
fix tests

hunterjm commented:
I've removed the DataUpdateCoordinator and added the websocket event implementation. This PR is now blocked by https://github.com/Kane610/aiounifi/pull/45.

hunterjm committed:
remove useless test

Kane610 commented on homeassistant/components/unifi/manifest.json:
```diff
@@ -3,7 +3,7 @@
   "config_flow": true,
   "documentation": "https://www.home-assistant.io/integrations/unifi",
-  "requirements": ["aiounifi==23"],
+  "requirements": ["aiounifi==24"],
```
```suggestion
  "requirements": ["aiounifi==25"],
```

Kane610 commented on requirements_all.txt:
```diff
@@ -227,7 +227,7 @@ aioshelly==0.4.0
 
 # homeassistant.components.unifi
-aiounifi==23
+aiounifi==24
```
```suggestion
aiounifi==25
```

Kane610 commented on requirements_test_all.txt:
```diff
@@ -143,7 +143,7 @@ aioshelly==0.4.0
 
 # homeassistant.components.unifi
-aiounifi==23
+aiounifi==24
```
```suggestion
aiounifi==25
```

Kane610 reviewed (approved):
Beautiful

hunterjm committed:
bump aiounifi

hunterjm commented:
LGTM!

Kane610 commented:
@MartinHjelmare something you want to add here or can I merge it?

MartinHjelmare commented on homeassistant/components/unifi/switch.py:
```diff
@@ -284,3 +314,61 @@ async def options_updated(self) -> None:
+            "name": "Deep Packet Inspection Restrictions",
+            "manufacturer": ATTR_MANUFACTURER,
+            "model": "Deep Packet Inspection",
+            "entry_type": "service",
```
Isn't this entity representing a feature of the unifi network switch? Then the device info should return info for the unifi network switch.

hunterjm replied:
It’s representing info from the controller, or second best the router. The problem is that we can’t really determine which router is on the device and updating the rules on the physical devices is done behind the scenes by the controller.

I could rename this service “UniFi Controller”. It’s still more of a service than a device, but then we can group all settings related configurations under this.

MartinHjelmare replied:
Just to be clear, the device info should represent the thing the provides the entity feature. It should not represent the entity feature itself.

What is it that provides this entity feature? Are we saying that it's a service on the controller that provides this feature or can we just say that it's the controller that provides the feature?

hunterjm replied:
The controller is software, not a physical device. It can be installed as software on any machine, though newer equipment like the Dream Machine has the controller software installed on the routing device. There is also the "cloud key" which is older separate hardware that had the controller pre-installed. Therefore it's a service on the controller, not a physical device itself.

MartinHjelmare replied:
Ok, good! Sounds good to name it UniFi Controller.

hunterjm committed:
rename device to UniFi Controller per PR feedback
